src/Database/LSMTree/Internal/TempRegistry.hs

```diff
@@ -0,0 +1,202 @@
+    resourceRelease :: (m ())
+  }
+
+{-# SPECIALISE unsafeNewTempRegistry :: IO (TempRegistry IO) #-}
+-- | This is considered unsafe, because one should properly 'bracket' this
+-- function. Example:
+--
+-- @
+--  generalBracket unsafeNewTempRegistry unsafeReleaseTempRegistry
+-- @
+unsafeNewTempRegistry :: MonadMVar m => m (TempRegistry m)
+unsafeNewTempRegistry = TempRegistry <$> newMVar (TempRegistryState Map.empty Map.empty (ResourceId 0))
+
+{-# SPECIALISE unsafeReleaseTempRegistry :: TempRegistry IO -> ExitCase a -> IO () #-}
+-- | See 'unsafeNewTempRegistry'.
+unsafeReleaseTempRegistry :: MonadMVar m => TempRegistry m -> ExitCase a -> m ()
+unsafeReleaseTempRegistry reg ec = case ec of
+    ExitCaseSuccess{} -> mapM_ resourceRelease . tempFreed     =<< takeMVar (tempRegistryState reg)
+    _                 -> mapM_ resourceRelease . tempAllocated =<< takeMVar (tempRegistryState reg)
+
+
+{-# SPECIALISE allocateTemp :: TempRegistry IO -> IO a -> (a -> IO ()) -> IO a #-}
+-- | Temporarily allocate a resource.
+--
+-- This runs the @acquire@ function with async exceptions masked to ensure that
+-- acquired resources are always put into the registry. However, note that in
+-- general the following two expressions are not equivalent:
+--
+-- @
+--   allocateTemp reg acquire free
+--   acquire >>= \x -> allocateTemp reg free (pure x)
+-- @
+--
+-- Assuming that @acquire@ is not already exception safe, it is /not/
+-- exception-safe to pass the result of @acquire@ to @allocateTemp@: an async
+-- exception could be thrown in between @acquire@ and @allocateTemp@, which
+-- leaks resources.
+allocateTemp :: (MonadMask m, MonadMVar m) =>
+     TempRegistry m
+  -> m a
+  -> (a -> m ())
+  -> m a
+allocateTemp reg acquire free = mustBeRight <$> allocateEitherTemp reg (fmap Right acquire) free
+  where
+    mustBeRight :: Either Void a -> a
+    mustBeRight (Left  v) = absurd v
+    mustBeRight (Right a) = a
+
+{-# SPECIALISE allocateMaybeTemp :: TempRegistry IO -> IO (Maybe a) -> (a -> IO ()) -> IO (Maybe a) #-}
+-- | Like 'allocateTemp', but for resources that might fail to be acquired.
+allocateMaybeTemp ::
+     (MonadMask m, MonadMVar m)
+  => TempRegistry m
+  -> m (Maybe a)
+  -> (a -> m ())
+  -> m (Maybe a)
+allocateMaybeTemp reg acquire free = fromEither <$> allocateEitherTemp reg (toEither <$> acquire) free
+  where
+    toEither :: Maybe a -> Either () a
+    toEither Nothing  = Left ()
+    toEither (Just x) = Right x
+
+    fromEither :: Either () a -> Maybe a
+    fromEither (Left ()) = Nothing
+    fromEither (Right x) = Just x
+
+{-# SPECIALISE allocateEitherTemp :: TempRegistry IO -> IO (Either e a) -> (a -> IO ()) -> IO (Either e a) #-}
+-- | Like 'allocateTemp', but for resources that might fail to be acquired.
+allocateEitherTemp ::
+     (MonadMask m, MonadMVar m)
+  => TempRegistry m
+  -> m (Either e a)
+  -> (a -> m ())
+  -> m (Either e a)
+allocateEitherTemp reg acquire free =
+    mask_ $ do
+      eith <- acquire
```

Comment:
Do we want the whole `acquire` to be masked? Is there a problem with using [`mask`](https://hackage.haskell.org/package/base-4.19.0.0/docs/Control-Exception.html#v:mask) and restoring around `acquire`?

Reply:
Okay, resolved.